### PR TITLE
Clean up Rviz Motion Planning plugin, add tooltips

### PR DESCRIPTION
**Pull request by felixvd**
### Description

I had a go at making the plugin smaller. This version isn't perfect either, but what is? Here are the before and after images:

**Before:**
Screenshot from 2020-09-13 14-19-51 Screenshot from 2020-09-13 14-20-03

**After:**
Screenshot from 2020-09-13 14-21-00 Screenshot from 2020-09-13 14-24-09

Changes:  
- Added tooltips
- Reduced the minimum size of entry fields so they can partially hide numbers if required (users will resize the panel when they need to change the values). This should only affect the layout when the user resizes to _under_ the previous minimum size. The previous sizes should look the same as before.
- Reordered and renamed options in the Planning tab. Some of these had confused me when I started out. Most of these are just suggestions and up for discussion though. I just took the liberty while I was at it.

Notes:  
- I don't know if `allowReplanning` and `allowLooking` have any effect. I have a feeling that this is stub functionality added by Ioan, but that was never fully implemented. Because it seems like no one uses these checkboxes (and they are not mentioned in the tutorials afaik) I moved them down.
- There is a progress bar at the bottom that appears during planning, but not execution. It doesn't seem very useful. If it can't work during execution (I haven't looked into it), it might be better to remove it.
- I would actually like to turn off `Use Collision-Aware IK` by default. The robot looks like it freaks out whenever you move it into a collision, and you always lose your IK configuration. Opinions?
- [Qt SizePolicy documentation](https://doc.qt.io/archives/qt-4.8/qsizepolicy.html#Policy-enum) 

This should be cherry-picked to Kinetic and Melodic if possible.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Include a screenshot if changing a GUI
- [x] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

### CODE REVIEW

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:823]:
```suggestion
               <string>Allow approximate IK solutions. This might be helpful if an exact solution cannot be found, e.g. for an underactuated arm.</string>
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:845]: I guess this option refers to online monitoring of the executed trajectory and triggering replanning if a new collision pops up due to a changed scene. This should be clarified.
I'm not sure when this checkbox is considered at all. Only for `Plan & Execute`?
```suggestion
               <string>Trigger replanning if new collisions pop up during plan execution.</string>
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:870]: Again, referring to a tutorial might be helpful here.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:895]:
```suggestion
               <string>Allow control of start and end poses as well as triggering of planning (and execution?) from external programs via ROS topics.</string>
```

**Review by rhaschke — COMMENTED** [review]
I like these changes very much, particularly the reordering of checkboxes.
I'm also fine with reducing the minimal size of text fields - as long as the initial/preferred size still shows all content.
See inline comments for some suggestions regarding the tooltips.

**Comment by v4hn** [comment]
A small lobbying comment: Why not incorporate https://github.com/ros-planning/moveit/pull/2007 here @felixvd? :-)
I still think that pr should have been merged.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:552]: Adding links to tooltips does not make sense. An appropriate link could of course be added to the RViz tutorial(s) if it does not exist there yet.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:801]: Please keep the name as-is. "Reset IK on collision" is technically also wrong, because what exactly happens depends on the IK solver.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:823]:
```suggestion
               <string>Accept IK solutions, even if they deviate from the target pose. While technically incorrect, these solutions are often still intuitive when moving the interactive marker to unreachable targets.</string>
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:845]: >  I'm not sure when this checkbox is considered at all. Only for Plan & Execute?

Yes.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:870]:
```suggestion
               <string>Allow to reposition a sensor to resolve apparent collisions. A sensor needs to be set up and the new measurements have to affect the collision scene, otherwise it has no effect.</string>
```

**Comment by v4hn** [comment]
I would propose to move "Replanning" and "Sensor Positioning" all the way to the bottom of the list and explicitly add something like `Experimental Feature:` to the beginning of the resp. tooltips.
I know these have been experimental for years, but nobody works on them and most maintainers don't even know what they do. :-)

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:552]: I don't actually know how to set it up. We don't have a tutorial, and googling "moveit database" turns up [completely unrelated content](https://docs.ipswitch.com/MOVEit/Transfer2018/Help/Admin/en/index.htm#23428.htm) and the stale issue #807.  I have a strong feeling the database feature is just dead. Maybe something to discuss at the next meeting.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:801]: Ok. I had changed it because "Collision-aware" made me think "Why would I ignore collisions? Is this dangerous?" when I first saw it, but if it comes up elsewhere then I'm fine with it.

**Comment by felixvd** [comment]
Thanks for the helpful comments and history lessons :) Especially helpful because I also had to guess what some of the features were. I merged the suggestions in a way that should address the main concerns.

I'm happy to tweak the Path Constraints area while I'm at it, especially because it is useless without the extra work of setting up a database, but now that @JeroenDM has put so much work into #2273 , I would prefer to make it visible by making that part of the UI useful rather than removing it.

How about providing default constraints in the dropdown list, such as:  
- No end effector rotation
- X-rotation unconstrained, YZ fixed
- Y-rotation unconstrained, XZ fixed
- Z-rotation unconstrained, XY fixed  
- Keep EE link within sphere containing start and goal
- Keep EE link within 5 cm around linear path to goal

Currently the constraints are defined by passing a string to the move_group. For the default constraints, I would create the constraint messages in the plugin and hand those to the move_group. Additional constraint choices could still be populated from a database, even though I doubt that feature is used by anyone, as mentioned in the comment.

@JeroenDM If you write expressions for the default constraints given a start/goal state, I'll put it into the GUI.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: Ah I see, that's what happens. It looks like it only makes the plugin subscribe to the interactive marker topics with [this callback function](https://github.com/ros-planning/moveit/blob/master/moveit_ros/robot_interaction/src/robot_interaction.cpp#L674), but that doesn't trigger a plan or execution. I changed the text, please have a look.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: I guess [this](http://wiki.ros.org/warehouse_ros/Tutorials) comes closest? [Only this link comes up in our tutorials](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/benchmarking/benchmarking_tutorial.html), but it just "How to install dependencies" and not very helpful. I'm not sure if this feature is worth maintaining at all to be honest.

Comment (rhaschke) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: @v4hn, Is [this tutorial](https://ros-planning.github.io/moveit_tutorials/doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.html) you recently contributed related?
Skimming through this tutorial, it seems to be related to constraints. But the database created seems to be unrelated to the mongo database used e.g. for poses and scenes. Can you confirm this?

Comment (rhaschke) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: > Adding links to tooltips does not make sense.

I agree, tooltips cannot support links. However, referring to a tutorial would be very helpful I think.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: I never used this feature and it is completely unclear to me, which strategy is used to reposition the sensor.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: The checkbox triggers the following function, which also enables topics to trigger planning and execution:
https://github.com/ros-planning/moveit/blob/91e1f095c2e68187e19f975951011c20f0d8b51e/moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame.cpp#L257-L279

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:694]: To emphasize that maximal velocities can only be reduced:
```suggestion
                 <string>Factor (between 0 and 1) to scale down maximum joint velocities</string>
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: Same here.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:802]:
```suggestion
               <string>If checked, the IK solver tries to avoid collisions when searching for a start/end state set via the interactive marker.

This is usually achieved by random seeding, which can flip the robot configuration (e.g. elbow up/down). Note that motion planning is always collision-aware, regardless of this checkbox.</string>
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (rhaschke) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:849]: See my previous comment. Triggering planning and execution becomes also possible.
Not sure though, this is a useful feature. If I want to control the robot from an app, I would directly talk to the move_group node instead...

**Review by rhaschke — APPROVED** [review]
I fixed some types. Otherwise looks good to me.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:849]: I thought I changed that. I must have forgotten to save. Will fix tomorrow.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: The logic is quite straight-forward: If the cost of obstacles close to the path is too high, direct the sensor towards a cost source. It was used to get rid of kinect-based artefact in the octomap in the past and is implemented [here](https://github.com/ros-planning/moveit/blob/master/moveit_ros/planning/plan_execution/src/plan_with_sensing.cpp#L206).

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:849]: Users of this feature reported in github issues before. I agree this is not ideal, but I guess it can simplify User-Machine-Interaction pipelines.

**Comment by v4hn** [comment]
> I'm happy to tweak the Path Constraints area while I'm at it (it's annoying me that it is useless without the extra work of setting up a database) but now that @JeroenDM has put so much work into #2273 , I would prefer to make it visible by making that part of the UI useful rather than removing it.

Nobody ever talked about removing path constraints! The whole argument was against the (still unused) "goal tolerances" field.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: This is warehouse functionality and not related to the constraint approximation database

Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: I had no idea those two were different.

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (felixvd) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:849]: I figured it could be useful for some interface, even if it seems a bit strange. I updated the comment.

**Comment by felixvd** [comment]
You're right. I misread because the other PR is called "Goal Constraints". 

I removed the goal constraints and moved the Clear Octomap button down, since it shouldn't be in the Commands section to start with. I also made the distances in the Query group a little bit smaller to match the size to the Commands group:

Screenshot 2020-09-20 00:04:03

I won't do anything about the Path Constraints in this PR, but the offer stands @JeroenDM .

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui
Comment (v4hn) [reply moveit_ros/visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui]: the warehouse refers to the mongodb that can store a number of MoveIt-related datastructures.

The constraint approximation database stores a set of *joint states* that approximates the manifold of a specific constraint.

**Review by rhaschke — COMMENTED** [review]
I do consider "Clear Octomap" a command, though it's not a planning command. So, I suggest bringing this button back into the commands group.
Also, you suggested changing the default of "Collision-Aware IK" to unticked. That's not yet done.

**Comment by JeroenDM** [comment]
> @JeroenDM If you write expressions for the default constraints given a start/goal state, I'll put it into the GUI.

I like your list of example default constraints! Ideally, these could be implemented independently of which planner plugin is used. (So we can even reuse them for TrajOpt in the future.) But I'm not sure where these default implementations would go.

It would look like a factory method that takes in a start and goal state and spits out a [moveit_msgs/Constraints](http://docs.ros.org/latest/api/moveit_msgs/html/msg/Constraints.html) message.

But maybe we should move this discussion to a separate issue?

**Comment by felixvd** [comment]
> Also, you suggested changing the default of "Collision-Aware IK" to unticked. That's not yet done.

The checkbox is inactive for me when I add a Motion Planning display in a new `panda_moveit_config demo.launch` Rviz window, and when I open the ui file in QT Creator. Is there something else that needs to be changed?

> I do consider "Clear Octomap" a command, though it's not a planning command. So, I suggest bringing this button back into the commands group.

It is a command in the same way that  "Clear Planning Scene" is, but it seems very different from the ones that generate or execute a robot trajectory. Grouping them together doesn't seem intuitive, so I assumed the placement was due to the lack of space. I am not dead set on this layout, but I think separating it makes sense. Maybe someone else has an opinion?

> It would look like a factory method that takes in a start and goal state and spits out a moveit_msgs/Constraints message. But maybe we should move this discussion to a separate issue?

Yes, to both. Feel free to start a PR. I think you still have access to my fork.

**Comment by v4hn** [comment]
> Maybe someone else has an opinion?

In your first version you left some space above `Clear Octomap` as well. For me that suffices to make it clear they refer to different things.
The alternative layout makes the button much bigger than it should be, given that few people use it.
I would prefer it if we had another button right next to it to share the space, but I can't think of useful functionality to add there.

Imho, the more annoying issue with the button is that it is active *even when there is no octomap in the planning scene*.

**Comment by rhaschke** [comment]
> The checkbox was ticked in my last screenshot because I ran `moveit_resources_fanuc_moveit_config demo.launch`, which loads the plugin from its `.rviz` file, where the checkbox is saved that way.

That screenshot was indeed confusing to me.

**Comment by felixvd** [comment]
> In your first version you left some space above `Clear Octomap` as well. For me that suffices to make it clear they refer to different things.

That's the text/label where the planning result appears, so to me there's still no semantic separation. It's true that it's a space though.

How about a "Reset Goal/Start" button next to the Octomap button? Sometimes clicking the dropdown menus to select "Current" is tedious, especially if the GUI bugs out and garbles the options. 

> Imho, the more annoying issue with the button is that it is active _even when there is no octomap in the planning scene_.

Yes! Let me fix that before merging this.

**Comment by v4hn** [comment]
> How about a "Reset Goal/Start" button next to the Octomap button?
> Sometimes clicking the dropdown menus to select "Current" is tedious, especially if the GUI bugs out and garbles the options.

I'm undecided here. Guess I would use the button though instead of clicking "current" explicitly. :-)
Of course, if the GUI bugs out, we should fix that bug instead. ;D

> Yes! Let me fix that before merging this.

:+1:

**Comment by rhaschke** [comment]
> The more annoying issue with the button is that it is active _even when there is no octomap in the planning scene_.

I agree, but I think that should be tackled in a separate PR.

> How about a "Reset Goal/Start" button next to the Octomap button?

The semantics of this button isn't apparent to me: Do you want to reset to current?

> Sometimes clicking the dropdown menus to select "Current" is tedious, especially if the GUI bugs out and garbles the options.

Can you reproduce this somehow? Of course, would be great to fix any bug there... (in a separate PR as well).

**Comment by felixvd** [comment]
> The semantics of this button isn't apparent to me: Do you want to reset to current?

Yes, that's what I thought.

> Can you reproduce this somehow? Of course, would be great to fix any bug there... (in a separate PR as well).

Sadly I've never been able to reproduce it. It appears randomly from what I can tell, and restarting Rviz generally fixes the problem. It seems to be some issue with Qt/the GUI pipeline, not this plugin.

> I'm undecided here. Guess I would use the button though instead of clicking "current" explicitly. :-)

I'm also not convinced that it's necessary, but it might be convenient and save some clicks.

**Comment by felixvd** [comment]
The button enabling/disabling seemed to work when I tested it with `roslaunch moveit_tutorials detect_and_add_cylinder_collision_object_demo.launch` from [this tutorial](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/perception_pipeline/perception_pipeline_tutorial.html), but the octomap wasn't cleared when I pressed the button. Does it work for you?

**Comment by v4hn** [comment]
This might be an independent hiccup. I do remember an issue about the cleared (empty) octomap not showing up in RViz although it was actually cleared.
The tutorial keeps publishing the *same* octomap every 10 seconds, so it might be that you just never see a change but everything else is fine.
I agree with @rhaschke by now that it makes sense to look into this in a separate pull-request. Thanks for working on it! Could you migrate it to a separate PR? :-)

**Comment by felixvd** [comment]
Yes, but even when I cleared it, the Collision-Aware IK was considering the Octomap. Come to think of it, it's probably because the octomap is cleared via the service of the move group capability, but the planning scene inside the plugin isn't updated. Looking a bit closer, [this function](https://github.com/ros-planning/moveit/blob/master/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp#L532) doesn't remove the Octomap from the world, but only from the `octomap_monitor`(which is optional...?), and doesn't trigger a scene update.

**Comment by rhaschke** [comment]
I'm still in favour of moving the "clear octomap" button back into the commands group.

**Comment by felixvd** [comment]
Eh, I don't care to make that new button, so I moved the Clear Octomap one back and just reduced the spacings. If that looks too cramped, just revert the last commit.

Screenshot 2020-09-23 00:41:46

Does anyone know why the two columns match up in QtCreator but not in my Rviz window? That issue was already there for me on the master branch.

**Review by rhaschke — APPROVED** [review]

**Comment by rhaschke** [comment]
> Does anyone know why the bottoms of the two columns match up in QtCreator but not in my Rviz window?

Probably QtCreator and rviz use different fonts. The vertical space between commands and path constraints is a fixed spacer and thus doesn't adapt when resizing the window. Moving the spacers below those columns to a reasonable position, solves the problem:
image image

**Comment by v4hn** [comment]
Looks good to me. Thanks for refining minute details such as checkbox order, tab stops, et cetera as well!

**Pull request merged**
